### DISCUSSION TITLE
Unable to set credit card IBAN

### DISCUSSION THREAD

**Discussion by scottgrobinson** [firefly-iii/firefly-iii#6782] [Firefly III]
### Support guidelines

- [X] I've read the [support guidelines](https://github.com/firefly-iii/firefly-iii/blob/main/.github/support.md)

### I've found a bug and checked that ...

- [X] ... [the documentation](https://docs.firefly-iii.org/) does not mention anything about my problem
- [X] ... there are no open or closed issues that are related to my problem
- [ ] ... it's [definitely me, not you](https://github.com/firefly-iii/firefly-iii/blob/main/.github/its_you_not_me.md)

### Description

When using the importer (& Nordigen), my credit card IBANs appear as "552157******0123" 

However, I'm not able to add either the full credit card number, or the obfustracted credit card number shown above (& in the importer) to Firefly due to the error "This is not a valid IBAN" - This prevents the importer from automatically mapping my credit card, therefor the importer has to be run manually by myself at present.

Am I missing something, or is there a way around this check?

### Expected behaviour

I can enter the IBAN provided by Nordigen into Firefly to allow the automatic importer mapping to work.

### Debug information

Debug information generated at 2022-12-31 12:23:28 Europe/London for Firefly III version **5.7.16**.

| Scope | Version |
| --- | --- |
| Firefly III | 5.7.16 |
| Firefly III API | 1.5.6 |
| Build | 316, 25-12-2022 07:50:13 CET |
| DB version | 18 (exp. 18) |
| PHP | `8.1.13` |
| Host | `Linux` |

| System info | Value |
| --- | --- |
| System TZ | Europe/London |
| Browser TZ | Europe/London |
| App environment | local |
| App debug mode | false |
| App cache driver | redis |
| App logging | notice, stack |
| Display errors | Off |
| Error reporting | ALL errors |
| Interface | apache2handler |
| Default language | en_US |
| Default locale | equal |
| BCscale | 24 |
| DB drivers | mysql, pgsql, sqlite |
| Current driver | mysql |
| Trusted proxies (.env) | `*` |

| User info | Value |
| --- | --- |
| Login provider | eloquent  |
| User guard | web  |
| Headers | N/A, N/A  |
| Stateful domains |  |
| Session start | 2022-12-01 00:00:00 |
| Session end | 2022-12-31 23:59:59 |
| Session first | 2022-09-28 00:00:00 |
| User ID | 1 |
| User language | en_GB |
| User locale | en_GB |
| Attempt at "en_GB.utf8" | 'en_GB.utf8' |
| Attempt at "en_GB.UTF-8" | 'en_GB.UTF-8' |
| User agent | Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36 |

**Comment by JC5** [comment]
Credit card numbers are not IBAN's. Use the "account number" field to store it in Firefly III.

**Comment by scottgrobinson** [comment]
@JC5 Thank you, I had tried that without much luck though...

Credit Card Account, Full credit card numnber - No mapping
Credit Card Account, Obfustracted credit card number (matching what is shown coming from Nordigen in FIDI) - No mapping
Default Account, Full credit card numnber - No mapping
Default Account, Obfustracted credit card number (matching what is shown coming from Nordigen in FIDI) - No mapping

Is there any debugging you can suggest that I can dig into here? Otherwise I'll get some print statements out in the code.... :)

**Reply by JC5** [reply]
To be honest, I'm not entirely sure what the problem is. The UI of the data importer allows you to mix and match all accounts found in your Nordigen data with any account in Firefly III. If the software doesn't pick it up, you can always use that.

**Reply by scottgrobinson** [reply]
Problem is that I can't automate (I.E Use the importer HTTP post functionallity) any import that includes multiple credit cards.

In the importer (via Nordigen), It will map Nordigen IBAN to Firefly accounts with the same IBAN, but It's not mapping Nordigen IBAN for credit cards to Firefly accounts. Having a very breif look at the code, it looks like mergeNordigenAccountLists only supports merging where IBANs are used, not account numbers. I.E I can't have Firefly map the Nordigen IBAN that it's showing for my credit card to the account number that I've stored for a credit card.

I've found a "workaround" for this - If I add the IBAN that Nordigen returns to the account directly in the database, all is good. 
Seems like two "possible" solutions:
* Unrestrict the entry of IBANs in the UI. Whilst I know that "552157******1234" is not a valid IBAN, it's the IBAN that is returned by Nordigan!
* Add mapping of Noridgen IBAN > Firefly Account Number IF it doesn't find a matching IBAN number in the database.

**Reply by JC5** [reply]
But there is an option to map these accounts manually in the UI. The selection is saved to the JSON file. Then, you can use the JSON file.

**Reply by scottgrobinson** [reply]
I guess in that case we have a "different" problem. My JSON contains the following generated by the importer:
```
    "accounts": {
        "0642fe59-aece-406b-9271-09a0cf4b860c": 84,
        "10ec74f8-f341-40a0-bd78-6821922d1a48": 82,
        "21f6e06a-9604-4a4b-86e4-ccae99b7556c": 83,
        "50cc5bc5-3d5f-4090-aa48-d02f2d199bfc": 85
    },
```

But the UI doesn't actually "use" the mapping (Screenshots following another import using that JSON, where you can see the accounts aren't ticked, and the accounts aren't mapped)

<img width="1099" alt="Screenshot 2022-12-31 at 14 54 47" src="https://user-images.githubusercontent.com/207075/210140932-b8c68e16-1b3c-4693-b5f7-7bfe09f50cdf.png">
<img width="1082" alt="Screenshot 2022-12-31 at 14 54 55" src="https://user-images.githubusercontent.com/207075/210140931-7ccb32bf-b0dc-462d-b3b9-da4db2bee74c.png">

**Reply by JC5** [reply]
Ah that could be a bug yes. Thanks for letting me know!

**Reply by JC5** [reply]
The next release of the data importer will match account numbers AND ibans, and it will correctly pre-select accounts.

**Comment by scottgrobinson** [comment]
Did you read my mind? I literally opened GitHub about 5 minutes ago to look
at this issue and a way around it…… thank you!!

On Sat, 14 Jan 2023 at 19:26, James Cole ***@***.***> wrote:

> The next release of the data importer will match account numbers AND
> ibans, and it will correctly pre-select accounts.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orgs/firefly-iii/discussions/6782#discussioncomment-4686609>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABSRYYBAIHKCMKZJD23WLTWSL4V5ANCNFSM6AAAAAATNQ2YM4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***
> com>
>
-- 
--
scott robinson

**Reply by JC5** [reply]
Glad to be of service, proud to hear you enjoy Firefly III!
